dev_tools/packaging/isolated_packages_test.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import subprocess
+from unittest import mock
+
+import pytest
+
+from dev_tools.modules import list_modules
+
+# point, that is then cloned to a separate folder for each test.
+PACKAGES = [
+    "-r",
+    "dev_tools/requirements/deps/pytest.txt",
+    "-r",
+    # one of the _compat_test.py tests uses flynt for testing metadata
+    "dev_tools/requirements/deps/flynt.txt",
+]
+
+
+@pytest.mark.slow
+# ensure that no cirq packages are on the PYTHONPATH, this is important, otherwise
+# the "isolation" fails and for example cirq-core would be on the PATH
+@mock.patch.dict(os.environ, clear='CIRQ_TESTING')
+@pytest.mark.parametrize('module', list_modules(), ids=[m.name for m in list_modules()])
+def test_isolated_packages(cloned_env, module):
+    env = cloned_env("isolated_packages", *PACKAGES)
+
+    if str(module.root) != "cirq-core":
+        result = subprocess.run(
+            f"{env}/bin/pip install ./{module.root}", capture_output=True, shell=True
+        )
+        assert result.returncode != 0, (
+            f"{module.name} should have failed to " f"install without cirq-core!"
```

Comment:
formatter artifact with two strings

Reply:
im confused by this assertion. shouldn't they all work and implicitly install cirq-core as part of their dependencies?

Reply:
Yes, that is true, however, when you are installing locally you have to list cirq-core as a folder as well, as pip will not be able to resolve `cirq-core==v0.12.0.dev` from PyPI. This then serves as a test for the dependency. I can add this clarification there as a comment.

Reply:
makes sense, comment would be helpful

Reply:
I thought one of the things that motivated this PR is that some of the modular packages were missing the `cirq-core` dependency. How are we testing that it is listed in the modules' requirements? Especially given the point above that the test infrastructure has to install cirq-core "manually".

Is this assertion supposed to be the test? If so, then it *definitely* needs a comment. Also other things could cause pip install to fail other than a missing cirq-core dep

Reply:
It is the test, as that is the user facing behavior we want for now (unless we figure out a way for local `pip install` / `setup.py install` to handle the local dependencies - related to https://github.com/quantumlib/Cirq/issues/4261). I agree that other things can cause pip install to fail, so it is starting to get a bit awkward to test the dependency this way. I completely agree that at minimum it would benefit from extensive comment explaining it. Alternatively I can add an explicit test for requirements (which should be easy).

Reply:
Went with an explicit test instead, it's also better as it explicitly tests the versioning logic too - the version of the cirq-core dependency should be the same as the module's version.